File: solutions/javascript/01_get_high_scores.js
const scores = require("./scores.json");
const benchmark = require("../utils/benchmark");

const getResults = () => {
  scores
    .filter((s) => s.active)
    .map((s) => {
      return {
        name: s.name,
        score: s.wins * 5 + s.draws * 2,
      };
    })
    .sort((a, b) => {
      if (a.score !== b.score) return b.score - a.score;
      const [afirstName, alastName] = a.name.split(" ");
      const [bfirstName, blastName] = b.name.split(" ");
      if (alastName < blastName) return -1;
      if (alastName > blastName) return 1;
      if (afirstName < bfirstName) return -1;
      if (afirstName > bfirstName) return 1;
      return 0;
    })

    .slice(0, 3);
};

benchmark(getResults, "getResults");

const getResultsFast = () => {
  const results = [];

  // a function to determine if the new result should push out the old
  // a represents the new result
  // b represents the existing result
  const isHigher = (a, b, result) => {
    if (!b) return true;

    if (result !== b.score) return b.score - result < 0;

    const [afirstName, alastName] = a.name.split(" ");
    const [bfirstName, blastName] = b.name.split(" ");

    if (alastName < blastName) return true;
    if (alastName > blastName) return false;
    if (afirstName < bfirstName) return true;
    if (afirstName > bfirstName) return false;
    return false;
  };

  for (const score of scores) {
    if (!score.active) continue;

    const result = score.wins * 5 + score.draws * 2;

    if (isHigher(score, results[0], result)) {
      results.unshift({ name: score.name, score: result });
      results.splice(3, 1);
      continue;
    }

    if (isHigher(score, results[1], result)) {
      results.splice(1, 0, { name: score.name, score: result });
      results.splice(3, 1);
      continue;
    }

    if (isHigher(score, results[2], result)) {
      results.splice(2, 0, { name: score.name, score: result });
      results.splice(3, 1);
    }
  }

  return results;
};

benchmark(getResultsFast, "getResultsFast");

File: solutions/go/01_get_high_scores.go
package main

import (
	"encoding/json"
	"fmt"
	"io"
	"os"
	"sort"
	"strings"
)

type Score struct {
	Name string `json:"name"`
	Wins int    `json:"wins"`
	Draws int   `json:"draws"`
	Active bool `json:"active"`
}

type HighScore struct {
	Name string
	Score int
}

type ByCombinedLogic []HighScore

func (a ByCombinedLogic) Len() int           { return len(a) }
func (a ByCombinedLogic) Less(i, j int) bool { 
	if (a[i].Score != a[j].Score) {
		return a[i].Score > a[j].Score
	}

	i_names := strings.Split(a[i].Name, " ")
	j_names := strings.Split(a[j].Name, " ")

	if (i_names[1] != j_names[1]) {
		return i_names[1] < j_names[1]
	}

	return i_names[0] < j_names[0]
}
func (a ByCombinedLogic) Swap(i, j int)      { a[i], a[j] = a[j], a[i] }

func Unmarshal(path string, data interface{}) error {
	jsonFile, err := os.Open(path)
	if err != nil {
		return err
	}

	defer jsonFile.Close()

	byteValue, _ := io.ReadAll(jsonFile)

	err = json.Unmarshal(byteValue, &data)
	if err != nil {
		return err
	}

	return nil
}

func main() {
	var scores []Score

	if err := Unmarshal("./scores.json", &scores); err != nil {
		panic(err)
	}

	var highScores []HighScore

	for _, score := range scores {
		if !score.Active {
			continue
		}

		result := (score.Wins * 5) + (score.Draws * 2)

		highScores = append(highScores, HighScore{
			Name: score.Name,
			Score: result,
		})
	}

	sort.Sort(ByCombinedLogic(highScores))

	topThree := []HighScore{
		highScores[0],
		highScores[1],
		highScores[2],
	}

	// prettyPrint, _ := json.MarshalIndent(topThree, "" ,"  ")
	// fmt.Println(string(prettyPrint))
	fmt.Println(topThree)
}